lib/vrl/parser/src/parser.lalrpop

```diff
@@ -0,0 +1,511 @@
+    Sp<AssignmentSingle>,
+    Sp<AssignmentInfallible>,
+};
+
+AssignmentSingle: Assignment = {
+    <target: Sp<AssignmentTarget>> "="
+        <expr: Box<Expr>> => Assignment::Single { target, expr },
+}
+
+AssignmentInfallible: Assignment = {
+    <ok: Sp<AssignmentTarget>> ","
+        <err: Sp<AssignmentTarget>> "="
+        <expr: Box<Expr>> => Assignment::Infallible{ ok, err, expr},
+}
+
+#[inline]
+AssignmentTarget: AssignmentTarget = {
+    "_" => AssignmentTarget::Noop,
+    "." <Path?> => AssignmentTarget::External(<>),
+    Ident Path? => AssignmentTarget::Internal(<>),
+};
+
+// -----------------------------------------------------------------------------
+// arithmetic
+// -----------------------------------------------------------------------------
+
+ArithmeticExpr: Node<Expr> = {
+    Sp<ErrorCoalesce>,
+};
+
+ErrorCoalesce: Expr = {
+    Op<ErrorCoalesce, "??", Logical>,
+    Logical,
+};
+
+Logical: Expr = {
+    Op<Logical, "||", Equal>,
+    Op<Logical, "&&", Equal>,
+    Equal,
+};
+
+Equal: Expr = {
+    Op<Equal, "!=", Compare>,
+    Op<Equal, "==", Compare>,
+    Compare,
+};
+
+Compare: Expr = {
+    Op<Compare, ">=", Add>,
+    Op<Compare, ">", Add>,
+    Op<Compare, "<=", Add>,
+    Op<Compare, "<", Add>,
+    Add,
+};
+
+Add: Expr = {
+    Op<Add, "+", Factor>,
+    Op<Add, "-", Factor>,
+    Factor,
+};
+
+Factor: Expr = {
+    Op<Factor, "%", Not>,
+    Op<Factor, "*", Not>,
+    Op<Factor, "/", Not>,
+    Not,
+};
+
+Not: Expr = {
+    <bang: Sp<"!">> <expr: Box<Sp<Not>>> => {
+        let span = span(bang.span().start(), expr.span().end());
+        let not = Unary::Not(Node::new(span, Not(bang.map(|_| ()), expr)));
+        Expr::Unary(Node::new(span, not))
+    },
+    Term,
+};
+
+Term: Expr = {
+    Sp<Literal> => Expr::Literal(<>),
+    Sp<Container> => Expr::Container(<>),
+    Sp<Query> => Expr::Query(<>),
+    Sp<FunctionCall> => Expr::FunctionCall(<>),
+    Sp<Ident> => Expr::Variable(<>),
+};
+
+// -----------------------------------------------------------------------------
+// query
+// -----------------------------------------------------------------------------
+
+pub Query: Query = {
+    LQuery <Sp<".">> RQuery => {
+        let span = <>.span();
+        let target = Node::new(span, QueryTarget::External);
+        let path = Node::new(span, Path(vec![]));
+
+        Query { target, path }
+    },
+    LQuery <target: Sp<QueryTarget>> <path: Sp<Path>> RQuery => Query { target, path },
+};
+
+QueryTarget: QueryTarget = {
+    Ident => QueryTarget::Internal(<>),
+    "." => QueryTarget::External,
+    FunctionCall => QueryTarget::FunctionCall(<>),
+    Container => QueryTarget::Container(<>),
+};
+
+// -----------------------------------------------------------------------------
+// path
+// -----------------------------------------------------------------------------
+
+Path: Path = PathSegment+ => Path(<>);
+
+PathSegment: PathSegment = {
+    "."? <Field> => PathSegment::Field(<>),
```

Comment:
Why is the leading `.` in the segment optional?

Reply:
It has to do with how we handle "queries".

This is a query: `.foo` which queries the "external target" (e.g. the event).

Here's another query: `foo.bar` which queries the local variable `foo`.

We store a query as a combination of "target" and "path". In the first case, the starting `.` is consumed and used to set the target to "external", then the next piece "foo" is used as the field name. In the second case `foo` is consumed and used to set the target to "variable foo", then the next piece ".bar" is used as the field name.

Maybe there are other ways to represent this, I didn't dig too deep, but I didn't see any way to match a token without actually consuming it, which is why I made the change here to also allow a path segment that doesn't start with the dot.